src/main/java/uk/gov/companieshouse/overseasentitiesapi/interceptor/UserAuthenticationInterceptor.java

```diff
@@ -74,4 +102,83 @@ public boolean preHandle(@NonNull HttpServletRequest request, @NonNull HttpServl
+            authInfoMap.put("roe_update_journey_company_numbers_match", doCompanyNumbersMatch);
+            if (doCompanyNumbersMatch) {
+                ApiLogger.debugContext(reqId, "UserAuthenticationInterceptor authorised with company_oe_annual_update=create permission",
+                        authInfoMap);
+                return true;
+            } else {
+                ApiLogger.errorContext(reqId, "UserAuthenticationInterceptor unauthorised for ROE Update Journey as Company Numbers inScope and Transaction do not match", null, authInfoMap);
+                return false;
+            }
+        } else {
+            ApiLogger.debugContext(reqId, "UserAuthenticationInterceptor unauthorised for ROE Update Journey as Company Number is not a valid OE",
+                    authInfoMap);
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return false;
+        }
+    }
+
+    private boolean isRequestRoeUpdate(String companyNumberInScope) {
+        // Further check for OE number is done in doRoeUpdateJourneyChecks method
+        return StringUtils.isNotEmpty(companyNumberInScope);
+    }
+
+    private String getCompanyNumberInScope (HttpServletRequest request) {
+        final Map<String, List<String>> privileges = getERICTokenPermissions(request);
+        if (privileges.containsKey(COMPANY_NUMBER_KEY)) {
+            return privileges.get(COMPANY_NUMBER_KEY).get(0);
+        }
+        return null;
+    }
+
+    private boolean doCompanyNumbersMatch(HttpServletRequest request, HttpServletResponse response, String companyNumberInScope, String transactionId) {
+        var companyNumberInTransactionOptional = getCompanyNumberInTransaction(request, response, transactionId);
+        if (companyNumberInTransactionOptional.isEmpty()) {
+            return false;
+        }
+        Optional<String> companyNumberInTransaction = companyNumberInTransactionOptional.get();
+        if (companyNumberInTransaction.isPresent() && companyNumberInTransaction.get().equalsIgnoreCase(companyNumberInScope)) {
+            return true;
+        }
+        response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+        return false;
+    }
+
+    private Map<String, List<String>> getERICTokenPermissions(HttpServletRequest request) {
+        String tokenPermissionsHeader = request.getHeader(ERIC_AUTHORISED_TOKEN_PERMISSIONS);
+        Map<String, List<String>> permissions = new HashMap<>();
+        if (tokenPermissionsHeader != null) {
+            for (String pair : tokenPermissionsHeader.split(" ")) {
+                String[] parts = pair.split("=");
+                permissions.put(parts[0], Arrays.asList(parts[1].split(",")));
+            }
+        }
+        return permissions;
+    }
+
+    private Optional<Optional<String>> getCompanyNumberInTransaction(HttpServletRequest request, HttpServletResponse response, String transactionId) {
```

Comment:
This method signature is now a little over-complicated. I've attempted to rework the logic in this [PR](https://github.com/companieshouse/overseas-entities-api/pull/303), which we can merge into this one, if you're happy.

Reply:
With the help from @AndreKramer7654 I've reworked the logic to simplify please review ...